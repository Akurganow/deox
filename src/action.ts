export type Action<
  Type extends string,
  Payload = undefined,
  Meta = undefined
> = Payload extends undefined
  ? (Meta extends undefined ? { type: Type } : { type: Type; meta: Meta })
  : (Payload extends Error
      ? (Meta extends undefined
          ? { type: Type; payload: Payload; error: true }
          : { type: Type; payload: Payload; meta: Meta; error: true })
      : (Meta extends undefined
          ? { type: Type; payload: Payload }
          : { type: Type; payload: Payload; meta: Meta }))
export type AnyAction = Action<string>

/**
 * Minimal (type-only) action factory
 * @example
 * const clearTodos = action('[Todo] truncate');
 */
export function action<Type extends string>(type: Type): Action<Type>
/**
 * Action with error factory
 * @example
 * const fetchTodosRejected = (payload: Error) => action('[Todo] fetch rejected', payload);
 */
export function action<Type extends string, Payload extends Error>(
  type: Type,
  payload: Payload
): Action<Type, Payload>
/**
 * Action with non-error payload factory
 * @example
 * const addTodo = ({ name, completed = false }: Todo) => action('[Todo] add', { name, completed });
 */
export function action<Type extends string, Payload>(
  type: Type,
  payload: Payload
): Action<Type, Payload>
/**
 * Action with error and meta factory
 * @example
 * const fetchTodosRejected = (payload: Error, meta?: Meta) => action('[Todo] fetch rejected', payload, meta);
 */
export function action<Type extends string, Payload extends Error, Meta>(
  type: Type,
  payload: Payload,
  meta: Meta
): Action<Type, Payload, Meta>
/**
 * Action with payload and meta factory
 * @example
 * const addTodo = ({ name, completed = false }: Todo, meta?: Meta) => action('[Todo] add', { name, completed }, meta);
 */
export function action<Type extends string, Payload, Meta>(
  type: Type,
  payload: Payload,
  meta: Meta
): Action<Type, Payload, Meta>

/**
 * Flux standard action factory
 * @example
 * const clearTodos = action('[Todo] truncate');
 * @example
 * const fetchTodosRejected = (payload: Error) => action('[Todo] fetch rejected', payload);
 * @example
 * const addTodo = ({ name, completed = false }: Todo) => action('[Todo] add', { name, completed });
 * @example
 * const fetchTodosRejected = (payload: Error, meta?: Meta) => action('[Todo] fetch rejected', payload, meta);
 * @example
 * const addTodo = ({ name, completed = false }: Todo, meta?: Meta) => action('[Todo] add', { name, completed }, meta);
 */
export function action<Type extends string, Payload, Meta>(
  type: Type,
  payload?: Payload,
  meta?: Meta
) {
  return {
    type,
    ...(payload ? { payload } : {}),
    ...(meta ? { meta } : {}),
    ...(payload instanceof Error ? { error: true } : {}),
  }
}
