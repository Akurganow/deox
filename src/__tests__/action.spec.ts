import { action } from '../action'

describe('action', () => {
  it.each`
    name                            | type                       | payload                      | meta
    ${'type-only'}                  | ${'[Todo] truncate'}       | ${undefined}                 | ${undefined}
    ${'error payload'}              | ${'[Todo] fetch rejected'} | ${new Error()}               | ${undefined}
    ${'error and meta'}             | ${'[Todo] fetch rejected'} | ${new Error()}               | ${{ status: 403 }}
    ${'non-error payload'}          | ${'[Todo] add'}            | ${{ name: `Read Robbin's` }} | ${undefined}
    ${'non-error payload and meta'} | ${'[Todo] add'}            | ${{ name: `Read Robbin's` }} | ${'Book is on bookshelf'}
  `('should create action with $name', ({ type, payload, meta }) => {
    expect(action(type, payload, meta)).toMatchSnapshot()
  })
})
