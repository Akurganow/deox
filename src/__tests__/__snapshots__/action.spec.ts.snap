// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action should create action with "error and meta" 1`] = `
Object {
  "error": true,
  "meta": Object {
    "status": 403,
  },
  "payload": [Error],
  "type": "[Todo] fetch rejected",
}
`;

exports[`action should create action with "error payload" 1`] = `
Object {
  "error": true,
  "payload": [Error],
  "type": "[Todo] fetch rejected",
}
`;

exports[`action should create action with "non-error payload and meta" 1`] = `
Object {
  "meta": "Book is on bookshelf",
  "payload": Object {
    "name": "Read Robbin's",
  },
  "type": "[Todo] add",
}
`;

exports[`action should create action with "non-error payload" 1`] = `
Object {
  "payload": Object {
    "name": "Read Robbin's",
  },
  "type": "[Todo] add",
}
`;

exports[`action should create action with "type-only" 1`] = `
Object {
  "type": "[Todo] truncate",
}
`;
